### PR TITLE
feat: Adding support of client authentication method.

### PR DESCRIPTION
**Pull request by BigTailWolf**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-auth-library-nodejs/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [X] Ensure the tests and linter pass
- [X] Code coverage does not decrease (if any source code was changed)
- [X] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: src/auth/oauth2client.ts [hunk 1]
```diff
@@ -68,6 +68,12 @@ export enum CertificateFormat {
   JWK = 'JWK',
 }
 
+export enum ClientAuthentication {
```
Comment (aeitzman) [inline src/auth/oauth2client.ts:71]: Can you add a comment explaining what this enum is for?

File: src/auth/oauth2client.ts [hunk 2]
```diff
@@ -475,6 +481,11 @@ export interface OAuth2ClientOptions extends AuthClientOptions {
    * The allowed OAuth2 token issuers.
    */
   issuers?: string[];
+  /**
+   * The client_authentication choice from baisc/post/none
```
Comment (aeitzman) [inline src/auth/oauth2client.ts:485]:
```suggestion
   * The client authentication type. Support values are basic, post, and none.
```

File: src/auth/oauth2client.ts [hunk 3]
```diff
@@ -475,6 +481,11 @@ export interface OAuth2ClientOptions extends AuthClientOptions {
...
+   * https://docs.authlib.org/en/latest/client/oauth2.html#client-authentication
```
Comment (aeitzman) [inline src/auth/oauth2client.ts:486]: Why is this linking to a specific python implementation? I think we should link to the rfc directly instead ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591#section-2))

File: src/auth/oauth2client.ts [hunk 4]
```diff
@@ -542,6 +554,8 @@ export class OAuth2Client extends AuthClient {
       oauth2IapPublicKeyUrl: 'https://www.gstatic.com/iap/verify/public_key',
       ...opts.endpoints,
     };
+    this.client_authentication =
+      opts.client_authentication || ClientAuthentication.ClientSecretPost;
```
Comment (aeitzman) [inline src/auth/oauth2client.ts:558]: Is this supposed to be post? The RFC says the default should be basic.

File: test/test.oauth2.ts [hunk 5]
```diff
@@ -1421,6 +1426,23 @@ describe('oauth2', () => {
       assert.strictEqual(params.client_id, 'overridden');
     });
 
+    it('getToken should use basic header auth if provided in options', async () => {
```
Comment (aeitzman) [inline test/test.oauth2.ts:1429]: can you add tests for the none and post type and a default case?

File: src/auth/oauth2client.ts [hunk 6]
```diff
@@ -660,20 +674,28 @@ export class OAuth2Client extends AuthClient {
     options: GetTokenOptions
   ): Promise<GetTokenResponse> {
     const url = this.endpoints.oauth2TokenUrl.toString();
-    const values = {
+    const headers: any = {'Content-Type': 'application/x-www-form-urlencoded'};
```
Comment (aeitzman) [inline src/auth/oauth2client.ts:677]: Don't need to use any here, can do something like this instead, same with below

https://github.com/googleapis/google-auth-library-nodejs/blob/7dbedff829be093dad0a2da32f0ab12862d473bc/src/auth/awsrequestsigner.ts#L142C5-L155C6

File: src/auth/oauth2client.ts (hunk 4 above)
Comment (BigTailWolf) [reply src/auth/oauth2client.ts:558]: The legacy behavior is post for oauth2 endpoint.

File: src/auth/oauth2client.ts (hunk 3 above)
Comment (BigTailWolf) [reply src/auth/oauth2client.ts:486]: Yeah, let me change into https://datatracker.ietf.org/doc/html/rfc7591#section-2

File: src/auth/oauth2client.ts (hunk 1 above)
Comment (BigTailWolf) [reply src/auth/oauth2client.ts:71]: Added

File: src/auth/oauth2client.ts (hunk 6 above)
Comment (BigTailWolf) [reply src/auth/oauth2client.ts:677]: Thanks for the example. When I made this change, I tried something else and none of them pass the generic type check then I left an `any`. Changed upon it.

File: src/auth/oauth2client.ts [hunk 7]
```diff
@@ -475,6 +481,11 @@ export interface OAuth2ClientOptions extends AuthClientOptions {
    * The allowed OAuth2 token issuers.
    */
   issuers?: string[];
+  /**
+   * The client authentication type. Support values are basic, post, and none.
```
Comment (aeitzman) [inline src/auth/oauth2client.ts:485]:
```suggestion
   * The client authentication type. Supported values are basic, post, and none.
```

File: test/test.oauth2.ts (hunk 5 above)
Comment (BigTailWolf) [reply test/test.oauth2.ts:1429]: Yeah, no problem.

File: src/auth/oauth2client.ts [hunk 8]
```diff
@@ -475,6 +485,12 @@ export interface OAuth2ClientOptions extends AuthClientOptions {
    * The allowed OAuth2 token issuers.
    */
   issuers?: string[];
+  /**
+   * The client authentication type. Supported values are basic, post, and none.
+   * If no type is provided, will default to post.
```
Comment (lsirac) [inline src/auth/oauth2client.ts:490]:
```suggestion
   * Defaults to post if not provided.
```

File: src/auth/oauth2client.ts [hunk 9]
```diff
@@ -491,6 +507,7 @@ export class OAuth2Client extends AuthClient {
   protected refreshTokenPromises = new Map<string, Promise<GetTokenResponse>>();
   readonly endpoints: Readonly<OAuth2ClientEndpoints>;
   readonly issuers: string[];
+  private client_authentication: ClientAuthentication;
```
Comment (lsirac) [inline src/auth/oauth2client.ts:510]: Looks like this should be readonly if we follow the existing convention

File: src/auth/oauth2client.ts [hunk 10]
```diff
@@ -660,20 +679,30 @@ export class OAuth2Client extends AuthClient {
...
   ): Promise<GetTokenResponse> {
     const url = this.endpoints.oauth2TokenUrl.toString();
-    const values = {
+    const headers: {[key: string]: string} = {
+      'Content-Type': 'application/x-www-form-urlencoded',
+    };
+    if (this.client_authentication === ClientAuthentication.ClientSecretBasic) {
+      const basic_auth =
+        'basic ' + btoa(`${this._clientId}:${this._clientSecret}`);
+      headers.Authorization = basic_auth;
```
Comment (lsirac) [inline src/auth/oauth2client.ts:688]:
```suggestion
      headers['Authorization'] = basic_auth;
```

File: src/auth/oauth2client.ts [hunk 11]
```diff
@@ -660,20 +679,30 @@ export class OAuth2Client extends AuthClient {
     options: GetTokenOptions
   ): Promise<GetTokenResponse> {
     const url = this.endpoints.oauth2TokenUrl.toString();
-    const values = {
+    const headers: {[key: string]: string} = {
```
Comment (lsirac) [inline src/auth/oauth2client.ts:682]:
```suggestion
    const headers: Headers = {
```

File: src/auth/oauth2client.ts [hunk 12]
```diff
@@ -660,20 +679,30 @@ export class OAuth2Client extends AuthClient {
...
+      'Content-Type': 'application/x-www-form-urlencoded',
+    };
+    if (this.client_authentication === ClientAuthentication.ClientSecretBasic) {
+      const basic_auth =
+        'basic ' + btoa(`${this._clientId}:${this._clientSecret}`);
+      headers.Authorization = basic_auth;
+    }
+    const values: {[key: string]: string | undefined} = {
```
Comment (lsirac) [inline src/auth/oauth2client.ts:690]:
```suggestion
    const values = {
```

File: src/auth/oauth2client.ts [hunk 13]
```diff
@@ -660,20 +679,30 @@ export class OAuth2Client extends AuthClient {
...
       code: options.code,
       client_id: options.client_id || this._clientId,
-      client_secret: this._clientSecret,
       redirect_uri: options.redirect_uri || this.redirectUri,
       grant_type: 'authorization_code',
       code_verifier: options.codeVerifier,
     };
+    if (this.client_authentication === ClientAuthentication.ClientSecretPost) {
+      values.client_secret = this._clientSecret;
```
Comment (lsirac) [inline src/auth/oauth2client.ts:698]:
```suggestion
   Object.assign(values, {client_secret: this._clientSecret});
```

File: test/test.oauth2.ts [hunk 14]
```diff
@@ -1421,6 +1426,68 @@ describe('oauth2', () => {
...
+        client_authentication: ClientAuthentication.ClientSecretBasic,
+      };
+      const oauth2client = new OAuth2Client(opts);
+      const _ = oauth2client.getToken({
+        code: 'code here',
+        client_id: CLIENT_ID,
+      });
+    });
+
+    it('getToken should not use basic header auth if provided none in options', async () => {
```
Comment (lsirac) [inline test/test.oauth2.ts:1458]: Where does the validation happen?

File: src/auth/oauth2client.ts [hunk 15]
```diff
@@ -68,6 +68,16 @@ export enum CertificateFormat {
   JWK = 'JWK',
 }
 
+/**
+ * The client authentication type. Support values are basic, post, and none.
```
Comment (aeitzman) [inline src/auth/oauth2client.ts:72]:
```suggestion
 * The client authentication type. Supported values are basic, post, and none.
```

File: src/auth/oauth2client.ts (hunk 9 above)
Comment (BigTailWolf) [reply src/auth/oauth2client.ts:510]: Done

File: test/test.oauth2.ts (hunk 14 above)
Comment (BigTailWolf) [reply test/test.oauth2.ts:1461]: Missed validation. Added.

File: test/test.oauth2.ts [hunk 16]
```diff
@@ -1421,6 +1427,70 @@ describe('oauth2', () => {
...
+      const opts = {
+        clientId: CLIENT_ID,
+        clientSecret: CLIENT_SECRET,
+        redirectUri: REDIRECT_URI,
+        endpoints: {
+          oauth2TokenUrl: 'mytokenurl',
+        },
+        client_authentication: ClientAuthentication.None,
+      };
+      const oauth2client = new OAuth2Client(opts);
```
Comment (lsirac) [inline test/test.oauth2.ts:1474]: You're not validating that its not using basic auth if client auth is set to None. This just validates the type, not the actual behavior

File: test/test.oauth2.ts [hunk 17]
```diff
@@ -1421,6 +1427,70 @@ describe('oauth2', () => {
...
+      const opts = {
+        clientId: CLIENT_ID,
+        clientSecret: CLIENT_SECRET,
+        redirectUri: REDIRECT_URI,
+        endpoints: {
+          oauth2TokenUrl: 'mytokenurl',
+        },
+      };
+      const oauth2client = new OAuth2Client(opts);
+      assert(
```
Comment (lsirac) [inline test/test.oauth2.ts:1488]: Same here

**Review by lsirac — CHANGES_REQUESTED** [review]

File: test/test.oauth2.ts (hunk 16 above)
Comment (BigTailWolf) [reply test/test.oauth2.ts:1474]: This part is validating the constructor behavior.
Technically we shouldn't use `None` auth either against OAuth2 portal or STS portal.

So shall here be a dummy token url and then throwing an error and validate that error as expected: an Auth should be used?

Comment (lsirac) [reply test/test.oauth2.ts:1477]: I think validating that there is no client auth in the request in the ways that are supported is enough

File: src/auth/oauth2client.ts [hunk 18]
```diff
@@ -68,6 +68,16 @@ export enum CertificateFormat {
...
 
+/**
+ * The client authentication type. Supported values are basic, post, and none.
+ * https://datatracker.ietf.org/doc/html/rfc7591#section-2
+ */
+export enum ClientAuthentication {
+  ClientSecretPost,
+  ClientSecretBasic,
+  None,
+}
```
Comment (danielbankhead) [inline src/auth/oauth2client.ts:79]:
```suggestion
export enum ClientAuthentication {
  ClientSecretPost = 'ClientSecretPost',
  ClientSecretBasic = 'ClientSecretBasic',
  None = 'None',
}
```

File: src/auth/oauth2client.ts [hunk 19]
```diff
@@ -491,6 +507,7 @@ export class OAuth2Client extends AuthClient {
   protected refreshTokenPromises = new Map<string, Promise<GetTokenResponse>>();
   readonly endpoints: Readonly<OAuth2ClientEndpoints>;
   readonly issuers: string[];
+  readonly client_authentication: ClientAuthentication;
```
Comment (danielbankhead) [inline src/auth/oauth2client.ts:510]:
```suggestion
  readonly clientAuthentication: ClientAuthentication;
```

File: src/auth/oauth2client.ts [hunk 20]
```diff
@@ -660,20 +679,34 @@ export class OAuth2Client extends AuthClient {
...
+      const basic_auth =
+        'Basic ' +
+        Buffer.from(`${this._clientId}:${this._clientSecret}`).toString(
+          'base64'
+        );
+      headers['Authorization'] = basic_auth;
+    } else if (
+      this.client_authentication === ClientAuthentication.ClientSecretPost
+    ) {
+      Object.assign(values, {client_secret: this._clientSecret});
```
Comment (danielbankhead) [inline src/auth/oauth2client.ts:702]:
```suggestion
      values.client_secret = this._clientSecret;
```

File: src/auth/oauth2client.ts [hunk 21]
```diff
@@ -660,20 +679,34 @@ export class OAuth2Client extends AuthClient {
...
+    }
     const res = await this.transporter.request<CredentialRequest>({
       ...OAuth2Client.RETRY_CONFIG,
       method: 'POST',
       url,
       data: querystring.stringify(values),
-      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
+      headers: headers,
```
Comment (danielbankhead) [inline src/auth/oauth2client.ts:709]:
```suggestion
      headers,
```

File: src/auth/oauth2client.ts [hunk 22]
```diff
@@ -660,20 +679,34 @@ export class OAuth2Client extends AuthClient {
     options: GetTokenOptions
   ): Promise<GetTokenResponse> {
     const url = this.endpoints.oauth2TokenUrl.toString();
+    const headers: Headers = {
+      'Content-Type': 'application/x-www-form-urlencoded',
+    };
     const values = {
```
Comment (danielbankhead) [inline src/auth/oauth2client.ts:685]: You can add `grant_type` to the `GetTokenOptions` type.

```suggestion
    const values: Omit<GetTokenOptions, 'codeVerifier'> = {
```

File: src/auth/oauth2client.ts [hunk 23]
```diff
@@ -660,20 +679,34 @@ export class OAuth2Client extends AuthClient {
...
       code: options.code,
       client_id: options.client_id || this._clientId,
-      client_secret: this._clientSecret,
       redirect_uri: options.redirect_uri || this.redirectUri,
       grant_type: 'authorization_code',
       code_verifier: options.codeVerifier,
```
Comment (danielbankhead) [inline src/auth/oauth2client.ts:690]:
```suggestion
      grant_type: options.grant_type || 'authorization_code',
      code_verifier: options.codeVerifier || options.code_verifier,
```

File: src/auth/oauth2client.ts [hunk 24]
```diff
@@ -68,6 +68,16 @@ export enum CertificateFormat {
...
+ */
+export enum ClientAuthentication {
+  ClientSecretPost,
+  ClientSecretBasic,
+  None,
+}
+
 export interface GetTokenOptions {
   code: string;
   codeVerifier?: string;
```
Comment (danielbankhead) [inline src/auth/oauth2client.ts:83]:
```suggestion
  codeVerifier?: string;
  /**
  * @alias GetTokenOptions['codeVerifier']
  **/
  code_verifier?: string;
  grant_type?: 'authorization_code';
```

File: test/test.oauth2.ts [hunk 25]
```diff
@@ -1421,6 +1427,88 @@ describe('oauth2', () => {
...
+        client_authentication: ClientAuthentication.ClientSecretBasic,
+      };
+      const oauth2client = new OAuth2Client(opts);
+      const res = await oauth2client.getToken({
+        code: 'code here',
+        client_id: CLIENT_ID,
+      });
+      scope.done();
+      assert(res.res);
+      assert(res.res.data.access_token === 'abc');
```
Comment (danielbankhead) [inline test/test.oauth2.ts:1461]: Improves the error message in the event of a failed assertion (prints expected/actual rather than `false`).

```suggestion
      assert.equal(res.res.data.access_token, 'abc');
```

File: test/test.oauth2.ts [hunk 26]
```diff
@@ -1421,6 +1427,88 @@ describe('oauth2', () => {
...
+        await oauth2client.getToken({
+          code: 'code here',
+          client_id: CLIENT_ID,
+        });
+      } catch (err) {
+        scope.done();
+        const e = err as GaxiosError;
+        assert(e);
+        assert.strictEqual(e.response!.status, 401);
+      }
```
Comment (danielbankhead) [inline test/test.oauth2.ts:1493]:
```suggestion
      try {
        await oauth2client.getToken({
          code: 'code here',
          client_id: CLIENT_ID,
        });
        throw new Error('Expected an error');
      } catch (err) {
        assert(err instanceof GaxiosError);
        assert.equal(err.response?.status, 401);
      } finally {
        scope.done();
      }
```

File: test/test.oauth2.ts [hunk 27]
```diff
@@ -1421,6 +1427,88 @@ describe('oauth2', () => {
...
+        redirectUri: REDIRECT_URI,
+        endpoints: {
+          oauth2TokenUrl: 'mytokenurl',
+        },
+      };
+      const oauth2client = new OAuth2Client(opts);
+      assert(
+        oauth2client.client_authentication ===
+          ClientAuthentication.ClientSecretPost
+      );
```
Comment (danielbankhead) [inline test/test.oauth2.ts:1509]:
```suggestion
      assert.equal(oauth2client.client_authentication, ClientAuthentication.ClientSecretPost);
```

**Review by danielbankhead — CHANGES_REQUESTED** [review]

File: src/auth/oauth2client.ts (hunk 20 above)
Comment (BigTailWolf) [reply src/auth/oauth2client.ts]: >  The compiler gives me error of `Property 'client_secret' does not exist on type 'Omit<GetTokenOptions, "codeVerifier">'.ts(2339)`

Comment (danielbankhead) [reply src/auth/oauth2client.ts:706]: We can add the type to `values`:

```
values: Omit<GetTokenOptions, 'codeVerifier'> & {client_secret: string}
```

Essentially, we don't want properties that are difficult to track down in the event of an error or general debugging.

**Review by danielbankhead — APPROVED** [review]

**Review by lsirac — APPROVED** [review]
